File: 01-Programming-Languages/PHP/Data-Structures/Heap/kth_largest.php
<?php

/*
Problem: Find Kth Largest Element
Description: Find the kth largest element in an array using a min-heap.

Approach:
- Use SplMinHeap to maintain the k largest elements seen so far
- For each element, push it; if heap size > k, pop the smallest
- The root of the heap is the kth largest

Time Complexity: O(n log k)
Space Complexity: O(k)

Example:
Input:  nums = [3, 2, 1, 5, 6, 4], k = 2
Output: 5
*/

function findKthLargest(array $nums, int $k): int {
    $heap = new SplMinHeap();
    foreach ($nums as $num) {
        $heap->insert($num);
        if ($heap->count() > $k) {
            $heap->extract();
        }
    }
    return $heap->top();
}

$nums = [3, 2, 1, 5, 6, 4];
$k = 2;
echo "{$k}th largest: " . findKthLargest($nums, $k) . "\n";

$nums = [3, 2, 3, 1, 2, 4, 5, 5, 6];
$k = 4;
echo "{$k}th largest: " . findKthLargest($nums, $k) . "\n";

File: 01-Programming-Languages/Ruby/Data-Structures/Heap/kth_largest.rb
=begin
Problem: Find Kth Largest Element
Description: Find the kth largest element in an array using a min-heap.

Approach:
- Maintain a min-heap of size k
- After processing all elements, the root is the kth largest

Time Complexity: O(n log k)
Space Complexity: O(k)

Example:
Input:  nums = [3, 2, 1, 5, 6, 4], k = 2
Output: 5
=end

def find_kth_largest(nums, k)
  heap = []
  nums.each do |num|
    heap.push(num)
    heap.sort!
    heap.shift if heap.length > k
  end
  heap.first
end

nums = [3, 2, 1, 5, 6, 4]
k = 2
puts "#{k}th largest: #{find_kth_largest(nums, k)}"

nums = [3, 2, 3, 1, 2, 4, 5, 5, 6]
k = 4
puts "#{k}th largest: #{find_kth_largest(nums, k)}"
